### PR TITLE
:bug: Misc. first fixes for 390, 391, 392, 393

### PR DESCRIPTION
**Pull request by ascheman**
Opening this as a draft for all of my work regarding #390, #391, #392, and #393 to enable further discussions.

### CODE REVIEW

File: provider/internal/java/util.go [hunk 1]
```diff
@@ -127,10 +127,10 @@ func decompile(ctx context.Context, log logr.Logger, filter decompileFilter, wor
 					continue
 				}
 				cmd := exec.CommandContext(
-					ctx, "java", "-jar", "/bin/fernflower.jar", job.inputPath, outputPathDir)
+					ctx, "java", "-jar", "/usr/local/bin/fernflower.jar", job.inputPath, outputPathDir)
```
Comment (pranavgaikwad) [inline provider/internal/java/util.go:130]: This will require us to fix [the Dockerfile in https://github.com/konveyor/java-analyzer-bundle so that ](https://github.com/konveyor/java-analyzer-bundle/blob/9cb5a19d87531487df378afbb37de4c60d16ae87/Dockerfile#L35), is his change an abosolute requirement?

Comment (ascheman) [reply provider/internal/java/util.go:130]: Correct, but it would make decompilation more portable as pointed out in b52929fb86eb7d455e9acf4af97b13cd7de40f09. I could not test with this path on MacOS, as /bin is write protected (for good reasons).

In general I would appreciate if such paths were configurable by file or environment. We could do this in the long run. 

Having said that, my inner monk ruffles up at the thought of having a jar file in the /bin folder (even if it's only a container, and even if the file is executable), as the /bin folder traditionally is the core folder for OS binaries. User add-ons should go to /usr/local or /opt.

Comment (jmle) [reply provider/internal/java/util.go:130]: I see that the main reason for this change is debugging. I'll send you some indications about how to debug from outside the container. I don't object to this change, but bare in mind that the analyzer will mostly be used from within the konveyor ecosystem in a containerized way, and that is the path we're mainly following.

Comment (ascheman) [reply provider/internal/java/util.go:130]: I fully understand that the end-user will mostly work with containers (which is perfectly fine for me) in order to avoid tons of local configurations, executables/tools etc. But to be honest, as a newbie development contributor, it is hard to follow all of these dependencies, and even harder to maintain them. 

Maybe for this change it is not much of a problem. But check out the change of Maven to 3.9 to avoid the compile goal (#393): 
* To get it running, 
  * you need to change the base image with the Java/Maven installations. 
  * Then build the analyzer-lsp image based on that. 
  * Finally build the Kantra image. 
* During runtime, 
  * `kantra` calls `konveyor-analyzer` (and some other tools), 
  * which call `mvn` at some point. 
If something goes wrong, it is hard to debug as the tools simply abort without leaving much trace to the root cause (perhaps I need to checkout the Jaeger integration?). I never saw that Maven failed (or even why) until I started the debugger.

To me it is much easier to run/debug all of this locally and have full control over each of the components.
Hence, I have the strong wish to be able to work locally with some (well-documented!) prerequisites as well as in containers. Setup should support both modes in my opinion.

BTW: Even when working completely in containers (as the end-user will do) it is essential to have a thorough documentation of all components, their workflows, all (intermediate) artifacts, and in particular how to observe their behaviour (logging/tracing/monitoring). Sorry, for repeating this over and over again ;-)

File: provider/internal/java/dependency.go [hunk 2]
```diff
@@ -138,9 +141,20 @@ func (p *javaServiceClient) GetDependenciesFallback(ctx context.Context, locatio
 		if *d.Version != "" {
 			if strings.Contains(*d.Version, "$") {
 				version := strings.TrimSuffix(strings.TrimPrefix(*d.Version, "${"), "}")
-				version = pom.Properties.Entries[version]
-				if version != "" {
+				p.log.V(10).Info("Searching for property in properties",
+					"property", version,
+					"properties", pom.Properties)
+				if pom.Properties == nil {
+					p.log.Info("Cannot resolve version property value as POM does not have properties",
```
Comment (shawn-hurley) [inline provider/internal/java/dependency.go:148]: very very small nit: what do we think of making the this `..."groupID", pomCo..., "artifactID", pomCor....)` 

Just wondering what folks think

File: provider/internal/java/dependency.go [hunk 3]
```diff
@@ -336,22 +350,47 @@ func (p *javaServiceClient) parseDepString(dep, localRepoPath string) (provider.
...
-	d.Version = parts[3]
-	d.Type = parts[4]
 
 	fp := filepath.Join(localRepoPath, strings.Replace(parts[0], ".", "/", -1), parts[1], d.Version, fmt.Sprintf("%v-%v.jar.sha1", parts[1], d.Version))
-	b, err := os.ReadFile(fp)
-	if err != nil {
-		// Log the error and continue with the next dependency.
-		p.log.V(5).Error(err, "error reading SHA hash file for dependency", "dep", d.Name)
-		// Set some default or empty resolved identifier for the dependency.
+	if _, err := os.Stat(fp); os.IsNotExist(err) {
```
Comment (shawn-hurley) [inline provider/internal/java/dependency.go:381]: Will ReadFile, give you the same error, or does it give a different error?

File: provider/internal/java/util.go (hunk 1 above)
Comment (shawn-hurley) [reply provider/internal/java/util.go:130]: While I agree in principle, the word support here is something to consider. 

Because the setup for the components is pretty complex, I worry about supporting everyone's local environments. This could be a long tail that would cost a lot of cycles, IMO. 

I would like to know if we can get an agreement on documentation and best effort to answer questions when testing locally, but there are no guarantees on time/fixes vs. in the container where there should be guarantees on this. Thoughts?

Comment (shawn-hurley) [reply provider/internal/java/util.go:130]: Oh, and in addition, I am plus on on moving things out of `/bin/` that seems like a good citizen thing to do.

File: provider/internal/java/dependency.go (hunk 3 above)
Comment (ascheman) [reply provider/internal/java/dependency.go:381]: ?? ReadFile returns an error if the file does not exist. Therefore, I consider it a good practice in programming to check whether the file exists, before trying to open it.

Comment (shawn-hurley) [reply provider/internal/java/dependency.go:381]: If that is the case, I would leave this logic as is. 

We get the same error, and it goes to the log explaining that the file does not exist. And we remove complexity to only have a single if/else block.

Comment (ascheman) [reply provider/internal/java/dependency.go:381]: Seems we have different opinions here. But may I remind you of the _good-citizen_ practice mentioned elsewhere? ;-)

File: provider/internal/java/util.go (hunk 1 above)
Comment (ascheman) [reply provider/internal/java/util.go:130]: Starting a smooth migration to better local executability / improving abilities to debug does not necessarily mean, that all possible paths must be fully supported. In this case, the change is invisible to almost everyone but makes life easier for someone developing outside of containers while not harming anyone else.

File: provider/internal/java/dependency.go (hunk 2 above)
Comment (ascheman) [reply provider/internal/java/dependency.go:148]: The pomCoordinate function was just a quick hack to cope with strings which are either set or "". Any better approach is highly appreciated.

File: provider/internal/java/dependency.go (hunk 3 above)
Comment (shawn-hurley) [reply provider/internal/java/dependency.go:381]: Please see the library function ReadFile:
 https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/os/file.go;drc=79a8997ac5592f11388d9f26ab3fff65ec876dbd;l=714

File: provider/internal/java/dependency.go (hunk 2 above)
Comment (shawn-hurley) [reply provider/internal/java/dependency.go:148]: Oh, I was not suggesting changing the helper function, I think that for this particular thing, it is a nice UX boost. I was discussing of making the log statement use the key, value pairs for groupID and artifactID instead of combining in a single key(POM). Just think it might be cleaner.

File: provider/internal/java/dependency.go (hunk 3 above)
Comment (fabianvf) [reply provider/internal/java/dependency.go:381]: I think in golang the typical practice would be to attempt to read the file and check the error for failure. 

As far as I can tell there are no differences between the approaches other than added handling complexity and an extra syscall if we check for existence first. The ReadFile function is built with the expectation that it may be called on non-existent files and has no unpleasant side-effects. 

Since we log the error anyway it will be clear the reason the file read failed without the extra check

Comment (pranavgaikwad) [reply provider/internal/java/dependency.go:381]: I think what shawn's pointing at is that - it's redundant to check for file exists, as the `err` variable in ReadFile will capture the same error itself.

Comment (pranavgaikwad) [reply provider/internal/java/dependency.go:381]: oops, sorry, fabian and I responded at the same time, +1

Comment (ascheman) [reply provider/internal/java/dependency.go:381]: Will not push this anymore. Nevertheless, an error is logged for missing .sha1 files which might not be in the local repo for my own modules. What bothers me here, is that this is logged as error. Errors imply that I (as a user/developer/ops person) have done something wrong. And there is no fault in this case. So for my own modules I would like to see the message logged in maximum as a warning, or even only info or lower (debug).

Comment (shawn-hurley) [reply provider/internal/java/dependency.go:381]: I think I agree with this and would plus one logging this as a V(5).Info while including the error as a key-value pair. You bring up a very good point about it being an error.

File: provider/internal/java/dependency.go [hunk 4]
```diff
@@ -336,22 +350,47 @@ func (p *javaServiceClient) parseDepString(dep, localRepoPath string) (provider.
...
+		p.log.V(5).Info("File does not exist", "file", fp)
 		d.ResolvedIdentifier = ""
 	} else {
-		d.ResolvedIdentifier = string(b)
+		b, err := os.ReadFile(fp)
+		if err != nil {
+			// Log the error and continue with the next dependency.
+			p.log.V(5).Error(err, "error reading SHA hash file for dependency", "dep", d.Name)
```
Comment (shawn-hurley) [inline provider/internal/java/dependency.go:388]: To be clear add an if statement here: `if os.IsNotExists(err) { p.log.V(5).Info("file does not exists", "file", fp, "err", err) } else { p.log.V(5).Error(err, "error reading SHA hash file for dependency", "dep", d.Name) } `
